### PR TITLE
[RF] Replace typeid with dynamic_cast in RooStatsUtils

### PR DESCRIPTION
**Pull request by guitargeek**
This addresses Jira issue [ROOT-4988](https://sft.its.cern.ch/jira/browse/ROOT-4899).

### CODE REVIEW

**Review by pcanal — APPROVED** [review]
Nice.

File: roofit/roostats/src/RooStatsUtils.cxx [hunk 1]
```diff
@@ -93,23 +93,19 @@ namespace RooStats {
...
-         RooAbsPdf *updf = dynamic_cast<RooAbsPdf *>(iter->Next());
-         assert(updf != 0);
-         delete iter;
-         FactorizePdf(observables, *updf, obsTerms, constraints);
-      } else if (id == typeid(RooSimultaneous)) {    //|| id == typeid(RooSimultaneousOpt)) {
-         RooSimultaneous *sim  = dynamic_cast<RooSimultaneous *>(&pdf);
+         auto iter = pdf.servers().begin();
+         assert(iter != pdf.servers().end());
+         assert(dynamic_cast<RooAbsPdf*>(*iter));
+         FactorizePdf(observables, static_cast<RooAbsPdf&>(**iter), obsTerms, constraints);
```
Comment (hageboeck) [inline roofit/roostats/src/RooStatsUtils.cxx:107]:
```diff
-      auto iter = pdf.servers().begin();
-      assert(iter != pdf.servers().end());
-      assert(dynamic_cast<RooAbsPdf*>(*iter));
-      FactorizePdf(observables, static_cast<RooAbsPdf&>(**iter), obsTerms, constraints);
+      assert(!pdf.servers().empty() && dynamic_cast<RooAbsPdf*>(pdf.servers().first()));
+      FactorizePdf(observables, static_cast<RooAbsPdf&>(*pdf.servers().first()), obsTerms, constraints);
```
I would even suggest to implement (== forward the call to the underlying vector) `RooSTLRefCountList::front()` and `back()`. The goal was to make it more STL-like anyway.

File: roofit/roostats/src/RooStatsUtils.cxx [hunk 2]
```diff
@@ -179,23 +173,24 @@ namespace RooStats {
...
 
-         TIterator *iter = pdf.serverIterator();
+         auto iter = pdf.servers().begin();
          // extract underlying pdf which is extended; first server is the pdf; second server is the number of events variable
-         RooAbsPdf *uPdf = dynamic_cast<RooAbsPdf *>(iter->Next());
-         RooAbsReal *extended_term = dynamic_cast<RooAbsReal *>(iter->Next());
-         assert(uPdf != NULL); assert(extended_term != NULL); assert(iter->Next() == NULL);
-         delete iter;
+         auto uPdf = dynamic_cast<RooAbsPdf *>(*(iter++));
+         auto extended_term = dynamic_cast<RooAbsReal *>(*(iter++));
```
Comment (hageboeck) [inline roofit/roostats/src/RooStatsUtils.cxx:181]: I didn't check, but I suggest to make them `const` if possible. RooFit doesn't do `const`ness very well, but we are in the process of improving that.

File: roofit/roostats/src/RooStatsUtils.cxx [hunk 3]
```diff
@@ -179,23 +173,24 @@ namespace RooStats {
...
          RooAbsPdf *newUPdf = StripConstraints(*uPdf, observables);
          if(newUPdf == NULL) return NULL; // only constraints in underlying pdf
          else return new RooExtendPdf(TString::Format("%s_unconstrained", pdf.GetName()).Data(),
             TString::Format("%s without constraints", pdf.GetTitle()).Data(), *newUPdf, *extended_term);
 
-      } else if (id == typeid(RooSimultaneous)) {    //|| id == typeid(RooSimultaneousOpt)) {
+      } else if (auto sim = dynamic_cast<RooSimultaneous *>(&pdf)) {  //|| dynamic_cast<RooSimultaneousOpt *>(&pdf)) {
 
-         RooSimultaneous *sim  = dynamic_cast<RooSimultaneous *>(&pdf); assert(sim != NULL);
+         assert(sim != nullptr);
```
Comment (hageboeck) [inline roofit/roostats/src/RooStatsUtils.cxx:193]: The `if` above will ensure that, won't it?

**Review by hageboeck — COMMENTED** [review]
Some post-merge ideas for future consideration.

**Pull request merged**